aws/resource_aws_macie_s3_bucket_association.go

```diff
@@ -0,0 +1,235 @@
+
+		Schema: map[string]*schema.Schema{
+			"bucket_name": {
+				Type:     schema.TypeString,
+				Required: true,
+				ForceNew: true,
+			},
+			"prefix": {
+				Type:     schema.TypeString,
+				Optional: true,
+				ForceNew: true,
+			},
+			"member_account_id": {
+				Type:         schema.TypeString,
+				Optional:     true,
+				ForceNew:     true,
+				ValidateFunc: validateAwsAccountId,
+			},
+			"classification_type": {
+				Type:     schema.TypeList,
+				Optional: true,
+				Computed: true,
+				MaxItems: 1,
+				Elem: &schema.Resource{
+					Schema: map[string]*schema.Schema{
+						"one_time": {
+							Type:     schema.TypeBool,
+							Optional: true,
+							Default:  false,
+						},
+					},
+				},
+			},
+		},
+	}
+}
+
+func resourceAwsMacieS3BucketAssociationCreate(d *schema.ResourceData, meta interface{}) error {
+	conn := meta.(*AWSClient).macieconn
+
+	req := &macie.AssociateS3ResourcesInput{
+		S3Resources: []*macie.S3ResourceClassification{
+			{
+				BucketName: aws.String(d.Get("bucket_name").(string)),
+			},
+		},
+	}
+	if v, ok := d.GetOk("member_account_id"); ok {
+		req.MemberAccountId = aws.String(v.(string))
+	}
+	if v, ok := d.GetOk("prefix"); ok {
+		req.S3Resources[0].Prefix = aws.String(v.(string))
+	}
+
+	ct := &macie.ClassificationType{
+		Continuous: aws.String(macie.S3ContinuousClassificationTypeFull),
+	}
+	ct.OneTime = aws.String(macieS3BucketAssociationOneTimeClassification(d))
+	req.S3Resources[0].ClassificationType = ct
+
+	log.Printf("[DEBUG] Creating Macie S3 bucket association: %#v", req)
+	resp, err := conn.AssociateS3Resources(req)
+	if err != nil {
+		return fmt.Errorf("Error creating Macie S3 bucket association: %s", err)
+	}
+	if len(resp.FailedS3Resources) > 0 {
+		return fmt.Errorf("Error creating Macie S3 bucket association: %s", resp.FailedS3Resources[0])
+	}
+
+	d.SetId(macieS3BucketAssociationId(d))
+	return resourceAwsMacieS3BucketAssociationRead(d, meta)
+}
+
+func resourceAwsMacieS3BucketAssociationRead(d *schema.ResourceData, meta interface{}) error {
+	conn := meta.(*AWSClient).macieconn
+
+	req := &macie.ListS3ResourcesInput{}
+	if v, ok := d.GetOk("member_account_id"); ok {
+		req.MemberAccountId = aws.String(v.(string))
+	}
+
+	bucketName := d.Get("bucket_name").(string)
+	prefix := d.Get("prefix")
+	var res *macie.S3ResourceClassification
+	for {
+		resp, err := conn.ListS3Resources(req)
+		if err != nil {
+			return fmt.Errorf("Error listing Macie S3 bucket associations: %s", err)
+		}
+
+		for _, v := range resp.S3Resources {
+			if aws.StringValue(v.BucketName) == bucketName && aws.StringValue(v.Prefix) == prefix {
+				res = v
+				break
+			}
+		}
+		if res != nil {
+			break
+		}
+
+		if resp.NextToken == nil {
+			break
+		}
+		req.NextToken = resp.NextToken
+	}
+
+	if res == nil {
+		log.Printf("[WARN] Macie S3 bucket association (%s) not found, removing from state", d.Id())
+		d.SetId("")
+		return nil
+	}
+
+	m := map[string]interface{}{}
```

Comment:
To prevent potential panics, we should perform a `nil` check for `res.ClassificationType` before dereferencing.

We also generally prefer to move this type of logic into a "flatten" function, e.g.

```go
func flattenMacieClassificationType(classificationType *macie.ClassificationType) []map[string]interface{} {
  if classificationType == nil {
    return []map[string]interface{}{}
  }
  m := map[string]interface{}{
    "one_time": false,
  }
  if aws.StringValue(res.ClassificationType.OneTime) == macie.S3OneTimeClassificationTypeFull {
    m["one_time"] = true
  }
  return []map[string]interface{}{m}
}
```

We should also provide a helpful error message when `d.Set()` fails, e.g.

```go
if err := d.Set("classification_type", flattenMacieClassificationType(res.ClassificationType)); err != nil {
  return fmt.Errorf("error setting classification_type: %s", err)
}
```

Reply:
👍